### PR TITLE
fix(server): fail job explicitly on unknown execution mode (#881)

### LINKED ISSUE #881: [Bug] Job executor silently accepts unknown modes and runs with partial behavior
`packages/server/src/repowise/server/job_executor.py:343`:

```python
mode = config.get("mode") or "sync"
is_full_resync = mode == "full_resync"
is_initial_index = mode == "initial_index"
is_index_only = mode == "index_only"
```

Any mode string that is not one of these four falls through every branch. The job still runs, still reports `completed`, and still advances `last_sync_commit`, but `generate_docs` is False and the incremental page-regen at line 420 (gated on `mode == "sync"`) is skipped. So an unknown mode silently produces a re-index with no documentation and a sync baseline that claims the repo is current.

This is what #837 was: webhooks enqueued `"incremental"`, which matches nothing. #848 fixed that caller. This issue is about the executor accepting the bad value in the first place, so the next typo or renamed mode fails the same silent way rather than loudly.

**Scope check.** `execute_job` is only reachable from three places, and the only modes that reach it are the four above:

- `routers/repos.py:705` (`initial_index`, `full_resync`)
- `routers/webhooks.py:57` (`sync`)
- `scheduler.py:162` (`sync`)

Other `{"mode": ...}` job rows exist but never reach the executor and should not be affected: `update_cmd/persistence.py:579` (`incremental`) and `upgrade_flow.py:243` (`upgrade`) are bookkeeping records so the web UI's "last synced" updates, and the latter is written with `status="completed"` directly.

**Fix direction:** validate `mode` against the known set at the top of `execute_job` and fail the job explicitly (recorded failure with a clear message) rather than running partial work. The default when `mode` is absent should stay `"sync"`.

**Worth noting while in there:** `routers/pages.py:121` enqueues a pending `{"mode": "single_page"}` job that nothing ever executes. `job_executor` has no `single_page` branch and no caller passes those job IDs to `execute_job`, so `POST /lookup/regenerate` returns `202 accepted` and appears to do nothing. That may want its own issue, but validation would have surfaced it.

### PR DESCRIPTION
**Pull request by akshatmalik-bruh**
Here is the exact filled-out content ready to copy and paste into your GitHub PR:

***

## Summary

- Validates job `mode` against `VALID_JOB_MODES = {"sync", "full_resync", "initial_index", "index_only"}` in `execute_job()`.
- Raises a `ValueError` on unknown modes so `execute_job()` explicitly fails the job and records the error message in the DB instead of silently running partial work and updating baseline commit state.
- Preserves `"sync"` as default fallback when `mode` is missing or empty.

## Related Issues

Fixes #881

## Test Plan

- [x] Tests pass (`pytest`) - Added unit tests `test_execute_job_fails_on_unknown_mode` and `test_execute_job_defaults_to_sync_when_mode_absent` in `tests/unit/server/test_job_executor.py`.
- [x] Lint passes (`ruff check .`)

## Checklist

- [x] My code follows the project's code style
- [x] I have added tests for new functionality
- [x] All existing tests still pass
- [x] I have updated documentation if needed (n/a — internal validation change)

***

### CODE REVIEW

File: packages/server/src/repowise/server/job_executor.py [hunk 1]
```diff
@@ -341,6 +345,12 @@ async def execute_job(
             wiki_style = _repo_wiki_style(repo, repo_path)
             config = json.loads(job.config_json) if job.config_json else {}
             mode = config.get("mode") or "sync"
```
Comment (swati510) [inline packages/server/src/repowise/server/job_executor.py]: This is a small robustness point rather than a real bug. The mode comes out of parsed JSON, so it is not guaranteed to be a string, and because `VALID_JOB_MODES` is a set, a non-empty list or dictionary would make the membership test on the next line raise an unhashable-type error instead of the helpful message you have written. The job would still be recorded as failed, but the message stored in the database would be confusing to whoever reads it later. Would you consider coercing the value when you read it, for example `mode = str(config.get("mode") or "sync")`, so that the friendly message is always the one that gets recorded?

File: packages/server/src/repowise/server/job_executor.py [hunk 2]
```diff
@@ -106,6 +106,10 @@ def _repo_wiki_style(repo: Any, repo_path: str) -> str:
     return resolve_style(style, repo_path=repo_path).name
 
 
+# Valid job execution modes handled by execute_job
```
Comment (swati510) [inline packages/server/src/repowise/server/job_executor.py]: Thank you for naming the accepted modes in one place. One thing that might be worth capturing here in a comment: `routers/pages.py` line 121 creates job rows with the mode "single_page", and those rows are never handed to `execute_job`, which is why they are not in this set. Without that note, a future reader who wires up the page regeneration endpoint will hit an "invalid job mode" failure and have to rediscover the history. This is only a comment suggestion, since the linked issue says the dormant `single_page` endpoint deserves its own ticket.

File: tests/unit/server/test_job_executor.py [hunk 3]
```diff
@@ -326,3 +326,68 @@ async def test_incremental_page_regen_passes_repo_path(tmp_path):
...
+    async with session_factory() as session:
+        updated_job = await get_generation_job(session, job_id)
+        assert updated_job is not None
+        assert updated_job.status == "failed"
+        assert "Invalid job mode 'incremental'" in (updated_job.error_message or "")
+
+
+@pytest.mark.asyncio
+async def test_execute_job_defaults_to_sync_when_mode_absent(session_factory, tmp_path):
+    """A missing or empty mode string defaults to 'sync' and proceeds without failing."""
```
Comment (swati510) [inline tests/unit/server/test_job_executor.py]: The description says "a missing or empty mode string", but the test only covers the missing case, since the configuration is an empty dictionary. The empty-string case is the one that actually exercises the `or "sync"` fallback, and it is the case most likely to break in a future refactor of that line. Could you either add a second case with `config={"mode": ""}` or adjust the description so that it matches what is being tested?

File: tests/unit/server/test_job_executor.py [hunk 4]
```diff
@@ -326,3 +326,68 @@ async def test_incremental_page_regen_passes_repo_path(tmp_path):
...
+        patch("repowise.server.job_executor.run_pipeline", run_pipeline_mock),
+        patch("repowise.server.job_executor.persist_pipeline_result", AsyncMock()),
+        patch(
+            "repowise.server.provider_config.get_chat_provider_instance",
+            side_effect=RuntimeError("no provider"),
+        ),
+    ):
+        await execute_job(job_id, app_state)
+
+    run_pipeline_mock.assert_awaited_once()
```
Comment (swati510) [inline tests/unit/server/test_job_executor.py]: This assertion shows that the pipeline ran, but it would pass for any accepted mode, so it does not quite prove that the fallback is "sync". Because the provider lookup is forced to fail in this test, the behaviour that is specific to "sync", namely the incremental page regeneration, never runs, so nothing here distinguishes "sync" from, say, "index_only". Checking the `generate_docs` keyword would not help either, since it is computed as false for "sync" and for "index_only" alike. What would distinguish them is that "sync" attempts the provider lookup while "index_only" skips that block entirely, so asserting that the patched `get_chat_provider_instance` was called would pin the mode down. Another option would be to let the patched provider return a mock client and then assert that the sync-only incremental page regeneration ran.

File: packages/server/src/repowise/server/job_executor.py (hunk 1 above)
Comment (akshatmalik-bruh) [reply packages/server/src/repowise/server/job_executor.py]: Done! Changed to mode = str(config.get("mode") or "sync") so any non-string JSON value (e.g. a list or dict) is coerced before the set membership test. This ensures the friendly ValueError with the invalid mode message is always the one recorded in the DB, rather than a confusing unhashable type error.

File: packages/server/src/repowise/server/job_executor.py (hunk 2 above)
Comment (akshatmalik-bruh) [reply packages/server/src/repowise/server/job_executor.py]: Added a comment explaining that "single_page" is intentionally excluded — routers/pages.py (line 121) creates those job rows but they are never dispatched to execute_job. Also noted the dormant endpoint deserves its own dedicated handling. Thanks for flagging — this will save future readers a confusing debugging session!

File: tests/unit/server/test_job_executor.py (hunk 3 above)
Comment (akshatmalik-bruh) [reply tests/unit/server/test_job_executor.py]: Good point! I've split this into two separate tests: test_execute_job_defaults_to_sync_when_mode_absent (covers config={}) and a new test_execute_job_defaults_to_sync_when_mode_empty_string (covers config={"mode": ""}) which directly exercises the or "sync" fallback branch. Also updated the docstring of the original test to only say "missing" instead of "missing or empty".

File: tests/unit/server/test_job_executor.py (hunk 4 above)
Comment (akshatmalik-bruh) [reply tests/unit/server/test_job_executor.py]: Great catch! I've captured the get_chat_provider_instance mock as get_provider_mock and added get_provider_mock.assert_called_once() after the run. Since "sync" always enters the provider lookup block while "index_only" skips it entirely, this assertion specifically pins the fallback mode down to "sync". Applied the same pattern to both the absent-mode and empty-string tests.

**Comment by akshatmalik-bruh** [comment]
Hi @swati510 @RaghavChamadiya  , just following up on this PR when you get a chance to take a look. Thanks!

**Comment by akshatmalik-bruh** [comment]
Resolved merge conflicts with main and verified tests pass. Ready for review whenever you get a chance!

**Comment by RaghavChamadiya** [comment]
Thanks for the quick turnaround on Swati's points. The coercion to `str(...)`, the split empty-string test, and the `get_provider_mock.assert_called_once()` assertion all address exactly what was raised, and that last one is a genuinely better test than what it replaced.

One blocking item, and it comes from the merge rather than from anything you wrote.

**`generate` and `single_page` are now dispatched modes, and this allowlist rejects them.** Since #881 was filed, `execute_job` gained a real branch for both:

```python
# job_executor.py:407 on this branch
if mode in ("generate", "single_page"):
    ...
    await _run_generate_job(...)
    return
```

Your merge commit pulled that in, but `VALID_JOB_MODES` (line 115) was not updated, and the validation at line 353 runs before the dispatch at 407. Both modes reach the executor for real:

- `routers/repos.py:667` builds `{"mode": "generate", ...}`, launched via `_launch_job_task`. That is `POST /repos/{id}/generate`, the HTTP `repowise generate`.
- `routers/pages.py:147` builds `{"mode": "single_page", ...}`, same path.

So after merge every generate and every per-page regenerate fails with `Invalid job mode 'generate'`. The endpoint that the issue described as dormant is live now.

To be clear about where this came from: the `single_page` note was Swati's suggestion on 2026-07-20 and it was correct then, and my own issue text said the same thing. The dispatch landed in between.

The fix:

1. Add `"generate"` and `"single_page"` to `VALID_JOB_MODES`.
2. Replace the exclusion note with the opposite, something like: these two are handled by the scoped-generation branch below, so validation has to run before it and accept them.
3. Add a test that a `{"mode": "generate"}` job still reaches `_run_generate_job`. That is the regression guard, and its absence is why CI stays green on this: nothing currently drives a generate job through `execute_job`.

One process note. The branch is 5 commits behind main again, and this is the second time the gap has hidden something. Could you rebase onto main rather than merging? A rebase would have put your validation line directly against the new dispatch branch and made the conflict visible instead of silently compatible. Worth a look at whether anything else moved under the PR while you were waiting.

Everything else here is ready. Fix the allowlist and I think this goes in.

**Comment by akshatmalik-bruh** [comment]
Thanks for catching that @RaghavChamadiya ! All items have been updated:

1. **Updated `VALID_JOB_MODES` Allowlist**: Added `"generate"` and `"single_page"` to `VALID_JOB_MODES` in `job_executor.py`.
2. **Updated Documentation**: Updated the inline comment above `VALID_JOB_MODES` to document that `"generate"` and `"single_page"` are valid modes dispatched to `_run_generate_job`.
3. **Regression Guard Test**: Added `test_execute_job_dispatches_generate_mode` in `tests/unit/server/test_job_executor.py` to ensure jobs with `mode="generate"` pass validation and reach `_run_generate_job`.
4. **Rebased onto Main**: Rebased the branch cleanly on top of latest `main`. All 23 unit tests in `test_job_executor.py` pass cleanly.

**Review by RaghavChamadiya — APPROVED** [review]
LGTM
